sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -0,0 +1,500 @@
+      // maxChunkSize is the Mongo partition size in MB
+      splitVectorCommand.append("maxChunkSize", desiredBundleSizeBytes / 1024 / 1024);
+      Document splitVectorCommandResult = mongoDatabase.runCommand(splitVectorCommand);
+      splitKeys = (ArrayList) splitVectorCommandResult.get("splitKeys");
+
+      return createSourceList(splitKeys);
+    }
+
+    /**
+     * Create a sources list. Each source in the list has a subset of keys to deal with (split).
+     *
+     * @param splitKeys List of split keys to create sources.
+     * @return The list of sources.
+     */
+    private List<BoundedSource> createSourceList(ArrayList splitKeys) {
+      List<BoundedSource> splitSourceList = new ArrayList<>();
+      if (splitKeys == null) {
+        splitSourceList.add(new BoundedMongoDbSource(uri, database, collection, filter, numSplits));
+      } else {
+        String lastID = null; // lower boundary of the first min split
+        int listIndex = 0;
+        for (final Object splitKey : splitKeys) {
+          String currentID = splitKey.toString();
+          String filterWithBoundaries;
+          if (filter != null && !filter.isEmpty()) {
+            if (listIndex == 0) {
+              filterWithBoundaries = "{ $and: [ {\"_id\":{$lte:ObjectId(\""
+                  + currentID + "\")}}, "
+                  + filter + " ]}";
+            } else if (listIndex == (splitKeys.size() - 1)) {
+              filterWithBoundaries = "{ $and: [ {\"_id\":{$gt:ObjectId(\"" + lastID
+                  + "\")," + "$lt:ObjectId(\"" + currentID
+                  + "\")}}, " + filter + " ]}";
+              splitSourceList.add(new BoundedMongoDbSource
+                  (uri, database, collection, filterWithBoundaries, numSplits));
+              filterWithBoundaries = "{ $and: [ {\"_id\":{$gt:ObjectId(\"" + currentID
+                  + "\")}}, " + filter + " ]}";
+            } else {
+              filterWithBoundaries = "{ $and: [ {\"_id\":{$gt:ObjectId(\"" + lastID
+                  + "\")," + "$lte:ObjectId(\"" + currentID + "\")}}, "
+                  + filter + " ]}";
+            }
+          } else {
+            if (listIndex == 0) {
+              filterWithBoundaries = "{\"_id\":{$lte:ObjectId(\"" + currentID + "\")}}";
+            } else if (listIndex == (splitKeys.size() - 1)) {
+              filterWithBoundaries = "{\"_id\":{$gt:ObjectId(\"" + lastID + "\"),"
+                  + "$lt:ObjectId(\"" + currentID + "\")}}";
+              splitSourceList.add(new BoundedMongoDbSource
+                  (uri, database, collection, filterWithBoundaries, numSplits));
+              filterWithBoundaries = "{\"_id\":{$gt:ObjectId(\"" + currentID + "\")}}";
+            } else {
+              filterWithBoundaries = "{\"_id\":{$gt:ObjectId(\"" + lastID + "\"),"
+                  + "$lte:ObjectId(\"" + currentID + "\")}}";
+            }
+          }
+          splitSourceList.add(new BoundedMongoDbSource
```

Comment:
I'd suggest to encapsulate the logic of this loop (converting a list of split keys into a list of filters) into a utility function, and separately invoke it to produce the actual BoundedMongoDbSource objects:

private static List<String> splitKeysToFilters(List<String> splitKeys, String additionalFilter) {
  ...
}

splitIntoBundles() {
  for (String shardFilter : splitKeysToFilters(splitKeys, filter)) {
    res.add(this.withFilter(shardFilter));
  }
}